CPU displacement calculation demo

oguzeroglu opened the pull request:
In this PR, a demo about applying the displacement map on the CPU is implemented. This method is inspired by the applyDisplacementMap functionality of Babylon.js. (Issue: #14698)

This might be useful since some materials do not support displacement maps (like MeshBasicMaterial) and some devices do not allow texture fetch inside the vertex shader so it may be possible for such devices to benefit displacement maps using the method implemented in this demo. Also, I find it more logical to apply the displacement map on the CPU before rendering as texture fetch operation inside shaders usually cause some sort of bottleneck.

SS:
![ekran resmi 2018-08-16 22 06 48](https://user-images.githubusercontent.com/15003836/44229423-b3e88700-a1a0-11e8-9b1c-4e41633a527f.png)

notoguz committed:
Geometry displacement example initial commits.

notoguz committed:
Modified HTML title

oguzeroglu committed:
CPU Displacement calculation demo.

Mugen87 commented:
For testing: https://rawgit.com/oguzeroglu/three.js/cpu_displacement_mapping/examples/webgl_geometry_displacement.html

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,291 @@
+
+				// map diffuse texture
+
+				var diffuseLoader = new THREE.TGALoader();
```
It should be sufficient to create a single instance of `TGALoader`  and reuse it.

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,291 @@
+				container = document.getElementById( 'container' );
+
+				camera = new THREE.PerspectiveCamera( 20, window.innerWidth / window.innerHeight, 1, 10000 );
+				camera.position.z = 100;
```
I think I would place the camera a bit further away so you can see the whole object from beginning. Besides, `OrbitControls` might be a better choice for camera controls.

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,291 @@
+
+			function displacementLoadEvent( textureData ) {
+
+				var geometry = new THREE.BoxGeometry( 50, 50, 50, 50, 50, 50 );
```
Please use `BufferGeometry` if possible.

oguzeroglu replied:
Do you mean the BoxBufferGeometry or the plain BufferGeometry?

Mugen87 replied:
I mean `BoxBufferGeometry`^^

Mugen87 commented:
This is optional but it would be really great if you have a configuration option (via dat.gui) to enable/disable the displacement. I think I would add two meshes (one with the displacement and a normal one) and then switch between both meshes (via the `visible` property).

mrdoob commented:
Wouldn't it be better to use pngs instead of tgas?

oguzeroglu commented:
@mrdoob AFAIK there's no difference between the TGA format and the PNG format if the GPU is concerned as all textures are ultimately converted to a format like DXT or BC5 when being sent to GPU (I might be wrong here). The only difference would be the amount of space they take in the hard disk. If that's what your concern is I can convert the TGA's to PNG's

Mugen87 commented:
> AFAIK there's no difference between the TGA format and the PNG format if the GPU is concerned as all textures are ultimately converted to a format like DXT or BC5 when being sent to GPU (I might be wrong here).

Texture compression does not happen not automatically. So referring to S3TC in this issue does not make sense.

oguzeroglu commented:
@Mugen87 Yes, you're right that's why I wrote I might be wrong here :) . I'm not sure about the format but I read somewhere that even though the image is a compressed image like JPEG, it does not matter for the GPU, it takes the same space as any other format when being uploaded to GPU if the texture is not properly compressed (like DXT).

Mugen87 commented:
JPEG is no real texture compression format. Since the size of a compressed pixel in JPEG varies, the GPU has to decompress the entire image for random access (which is necessary due to the massive parallel processing of the shaders). Texture compression formats like S3TC, ETC1, PVRTC and ASTC ensure that all texels allocate the same amount of data space within a texture. So there is no decompression necessary when uploading the texture.

oguzeroglu committed:
switched to buffer geometry

oguzeroglu committed:
gui integration

oguzeroglu commented:
@Mugen87 I've just commited the requested changes and also added the GUI. Also converted the textures from TGA to PNG as suggested by @mrdoob and switched to another texture which fits better for box-like geometries in my opinion.

I have to admit that porting the code to buffer geometry was a good idea, it made the code much clearer and easier.

oguzeroglu committed:
no need to set the camera position

notoguz committed:
no need for maxPolarAngle

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,288 @@
+			var controls;
+
+			var windowHalfX = window.innerWidth / 2;
+			var windowHalfY = window.innerHeight / 2;
```
These two variables are not necessary anymore.

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,288 @@
+
+				container = document.getElementById( 'container' );
+
+				camera = new THREE.PerspectiveCamera( 20, window.innerWidth / window.innerHeight, 1, 10000 );
```
I would place the camera in the allowed range. So maybe:
```js
camera.position.set( 0, 0, 300 );
```
This will also avoid that the camera looks down to your object which is currently the case. The current position is `(0, 250, 0)` which is computed by `OrbitControls` in order to ensure the camera is between `minDistance` and `maxDistance`.

Mugen87 commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,288 @@
+			function onWindowResize() {
+
+				windowHalfX = window.innerWidth / 2;
+				windowHalfY = window.innerHeight / 2;
```
Obsolete.

notoguz committed:
removed obsolete functions

oguzeroglu commented:
@Mugen87  Done!

Mugen87 commented:
Why have you removed `onWindowResize()` :wink: ? Only `windowHalfX` and `windowHalfY` are obsolete.

notoguz committed:
window resize fix.

oguzeroglu commented:
OH yeap, my bad. Fixed it.

oguzeroglu committed:
better asset loading method

WestLangley commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,283 @@
+			function mapEmissive ( textureData ) {
+
+				material.emissiveIntensity = 2;
+				material.emissive = new THREE.Color( 0xffffff );
```
don't forget this one! :)

WestLangley commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,318 @@
+
+				}
+
+				geometry.attributes.position.needsUpdate = true;
```
My comments are being lost for some reason... `needsUpdate` is not needed here.

WestLangley commented:
Have a look a `LoadingManager`. For example, #14741.

oguzeroglu committed:
removed redundant needsUpdate flag of buffer geometry

oguzeroglu committed:
integrated LoadingManager class

oguzeroglu commented:
@WestLangley done! I was not aware of the LoadingManager class, it made the code simpler. Thanks for the tip!

oguzeroglu committed:
consider texture transformation

oguzeroglu commented:
I did another modification in the last commit. Now instead of manually flipping the canvas, we simply transform the UV vector using the transformation matrix of the texture, therefore repeat/offset/wrapS/wrapT etc. properties are now considered when mapping the displacement texture.

WestLangley commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,266 @@
+
+			if ( ! Detector.webgl ) Detector.addGetWebGLMessage();
+
+			var TOTAL_ASSET_COUNT = 5;
```
This variable is not used.

oguzeroglu replied:
done

WestLangley commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,266 @@
+			if ( ! Detector.webgl ) Detector.addGetWebGLMessage();
+
+			var TOTAL_ASSET_COUNT = 5;
+			var loadedAssetCount = 0;
```
This variable is not used.

oguzeroglu replied:
done

WestLangley commented on examples/webgl_geometry_displacement.html:
```diff
@@ -0,0 +1,266 @@
+			var TOTAL_ASSET_COUNT = 5;
+			var loadedAssetCount = 0;
+
+			var materialParameters = { emissiveIntensity: 2, emissive: new THREE.Color( 0xffffff ) };
```
This is the 3rd time I have mentioned this. You do not need to instantiate an new instance to modify a color.
```
emissive: 0xffffff
```

oguzeroglu replied:
done

WestLangley commented:
Setting ambient light to `0xffffff` is not appropriate. It also defeats the purpose to use it as the only light source with `MeshPhongMaterial`.

You need to add a direct light source to your scene. If you do that, you will see there is something wrong with your normals.

My conjecture is your normal map is only appropriate for a non-displaced mesh.

oguzeroglu committed:
removed redundant variables.

oguzeroglu commented:
@WestLangley What do you mean by "something is wrong with your normals"? I think the problem here is that in this demo OrbitControls are used so the object itself is not moving but the camera is moving. When you put a directional light to a specific position without updating its position you'd see only one side of the cube is being illuminated. Maybe that's what you meant? I don't want to put 6 directional lights to the scene.

The reason I'm using the MeshPhongMaterial here is to have that sweet emissive intensity effect. I'm using hight quality textures here and the object looks good to me (Scifi vibes). I don't want to focus too much on non geometry related issues in THIS PR which is titled "CPU Displacement Calculation", I did this to demonstrate how displacement can be done on geometry level using Javascript (without GLSL).

Thanks for the feedback about the emissive parameter, I wasn't aware of that. I did not see the 2 other remarks of that issue though. I don't know why you wrote "this is the 3rd time I have mentioned this"

oguzeroglu committed:
removed normal mapping

Mugen87 commented:
@oguzeroglu Why have you closed the PR? It was just before the goal.

@WestLangley When creating review comments, you have to submit the entire review. Only then the comments become visible for others. Maybe that was the problem?

oguzeroglu commented:
> @oguzeroglu Why have you closed the PR? It was just before the goal.

I think I should do another PR directly to the geometry class, this PR became too material oriented (normals/lights etc.). I was expecting some feedback/discussion about the geometry and displacement stuff. At this point I don't have enough THREE.js knowledge to make a nice example.
